File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0733-2.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        int m = image.size();
        int n = image[0].size();
                
        // Did we visit this slot before?
        vector<vector<char>> color(m, vector<char>(n, 0));
        
        // <row, column>
        stack<pair<int, int>> S;
        S.push(make_pair(sr, sc));
        int initialColor = image[sr][sc];
        image[sr][sc] = newColor;
        color[sr][sc] = 1;

        // Expand by DFS
        while (!S.empty()){
            auto [r, c] = S.top(); S.pop();

            vector<pair<int, int>> neighs = {{r, c-1}, {r, c+1}, {r-1, c}, {r+1, c}};
            for (auto [nr, nc]: neighs){
                if (nr<0 || nr==m || nc<0 || nc == n) continue;
                if (color[nr][nc] == 0 && image[nr][nc] == initialColor){
                    image[nr][nc] = newColor;
                    color[nr][nc] = 1;
                    S.push(make_pair(nr, nc));
                } //end-if
            } // end-for
        } // end-while

        return image;
    } // end-floodFill
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0733-2.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        m = len(image)
        n = len(image[0])
        color = []
        for i in range(m): color.append([0]*n)
            
        # <row, column>
        S = []
        S.append([sr, sc])
        initialColor = image[sr][sc]
        image[sr][sc] = newColor
        color[sr][sc] = 1

        # Expand by DFS
        while len(S) > 0:
            r, c = S.pop()

            neighs = ((r, c-1), (r, c+1), (r-1, c), (r+1, c))
            for nr, nc in neighs:
                if nr<0 or nr==m or nc<0 or nc == n: continue
                if color[nr][nc] == 0 and image[nr][nc] == initialColor:
                    image[nr][nc] = newColor
                    color[nr][nc] = 1
                    S.append([nr, nc])
        return image
